source/common/router/scoped_rds.cc

```diff
@@ -101,70 +107,167 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
-      auto* thread_local_scoped_config =
-          const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
-      thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
-      return config;
-    });
+  for (const auto& scope_name : removed_resources) {
+    auto iter = scoped_route_map_.find(scope_name);
+    if (iter != scoped_route_map_.end()) {
+      ScopedRouteInfoConstSharedPtr to_be_deleted = iter->second;
+      scope_name_by_hash_.erase(iter->second->scopeKey().hash());
+      scoped_route_map_.erase(iter);
+      applyConfigUpdate(
+          [scope_name](
+              ConfigProvider::ConfigConstSharedPtr config) -> ConfigProvider::ConfigConstSharedPtr {
+            auto* thread_local_scoped_config =
+                const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
+            thread_local_scoped_config->removeRoutingScope(scope_name);
+            return config;
+          },
+          [to_be_deleted]() { /*to_be_deleted will be destructed in main thread.*/ });
+      any_applied = true;
+      ENVOY_LOG(debug, "srds: remove scoped route '{}'", scope_name);
+    }
   }
-
-  for (const auto& scoped_route : scoped_routes_to_remove) {
-    const std::string scoped_route_name = scoped_route.first;
-    ENVOY_LOG(debug, "srds: remove scoped route '{}'", scoped_route_name);
-    scoped_config_manager_.removeRoutingScope(scoped_route_name);
-    applyConfigUpdate([scoped_route_name](const ConfigProvider::ConfigConstSharedPtr& config)
-                          -> ConfigProvider::ConfigConstSharedPtr {
-      // In place update.
-      auto* thread_local_scoped_config =
-          const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
-      thread_local_scoped_config->removeRoutingScope(scoped_route_name);
-      return config;
-    });
+  ConfigSubscriptionCommonBase::onConfigUpdate();
+  if (any_applied) {
+    setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
   }
-
-  DeltaConfigSubscriptionInstance::onConfigUpdate();
-  setLastConfigInfo(absl::optional<LastConfigInfo>({absl::nullopt, version_info}));
   stats_.config_reload_.inc();
+  if (!exception_msgs.empty()) {
+    throw EnvoyException(fmt::format("Error adding/updating scoped route(s): {}",
+                                     StringUtil::join(exception_msgs, ", ")));
+  }
 }
 
+// TODO(stevenzzzz): see issue #7508, consider generalizing this function as it overlaps with
+// CdsApiImpl::onConfigUpdate.
+void ScopedRdsConfigSubscription::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const std::string& version_info) {
+  absl::flat_hash_map<std::string, envoy::api::v2::ScopedRouteConfiguration> scoped_routes;
+  absl::flat_hash_map<uint64_t, std::string> scope_name_by_key_hash;
+  for (const auto& resource_any : resources) {
+    // Throws (thus rejects all) on any error.
```

Comment:
added comment in the header file, this will also be addressed in fix for issue #7508.

Reply:
If you look at the history of CDS, we initially rejected all resources on any failure, but then we went back and plumbed in the ability to provide per-resource failure information. I think we might want that here.

Reply:
shall we follow up on this  issue in #7508?
I think it's a regression if xDS have different behaviors over different errors (e.g., fail on any failure for translation, and partial accept on validation error) in the SotW form onConfigUpdate. 

IMHO we should reject on any error for the SotW form API, and tolerate partial errors on the delta-form onConfigUpdate(added_resources, deleted_resources, version). But definitely we need to come to an consensus on this, esp when different XDS has different behavior already.

Reply:
I don't think this is possible, since there are Envoy operators who added the partial rejection explicitly to solve problems they were having, we can't regress them on CDS and we shouldn't be inconsistent here.

Reply:
OK, I see that we're just validating here, and all the acceptance exceptions happen when we call into the delta `onConfigUpdate`. This seems legit, maybe just add a comment explaining.